src/core/service.c

```diff
@@ -1439,6 +1439,34 @@ static bool service_exec_needs_notify_socket(Service *s, ExecFlags flags) {
         return s->notify_access != NOTIFY_NONE;
 }
 
+static Service *service_get_triggering_service(Service *s) {
+        assert(s);
+
+        /* Return the service which triggered service s, this means dependency
+         * types which include the UNIT_ATOM_PROPAGATE_EXIT_STATUS atom. */
+
+        Unit *other;
+
+        UNIT_FOREACH_DEPENDENCY(other, UNIT(s), UNIT_ATOM_PROPAGATE_EXIT_STATUS) {
+                if (other->type != UNIT_SERVICE)
+                        return NULL;
+
+                if (unit_has_dependency(other, UNIT_ATOM_ON_SUCCESS, UNIT(s))) {
+                        if (SERVICE(other)->result == SERVICE_SUCCESS &&
+                            unit_active_state(other) != UNIT_FAILED)
+                                return SERVICE(other);
+                }
+
+                if (unit_has_dependency(other, UNIT_ATOM_ON_FAILURE, UNIT(s))) {
+                        if (!IN_SET(SERVICE(other)->result, SERVICE_SUCCESS, SERVICE_SKIP_CONDITION) &&
+                            unit_active_state(other) == UNIT_FAILED)
+                                return SERVICE(other);
+                }
+        }
```

Comment:
so you are returning the first service you find here. There can be in theory multiple. i think it would be nice to log about cases like that (at least at debug level) and then return NULL if more than one is found.

Reply:
Yes, you are correct. I'll do as suggested.

Reply:
Yes, agreed this is indeed very limiting. I'm now thinking on how I can pass the triggering unit through to the triggered unit somehow, I'll likely push a WIP change for review at some point today.

Reply:
I pushed some changes to address this, though it's WIP. Looking for feedback on the validity of the new approach which just stashes a pointer to the triggering service. If this is ok I'll update the tests to cover more complicated scenarios.

Reply:
Docs and tests will get an update based on the outcome of this discussion.